adr/0008-phasing-out-config-classes.adoc

```diff
@@ -0,0 +1,95 @@
+
+We currently have two competing implementations to handle configuration properties in Quarkus extensions:
+
+- The legacy config class approach: it uses **classes** annotated with `@io.quarkus.runtime.annotations.ConfigRoot` and **fields** annotated with `@io.quarkus.runtime.annotations.ConfigItem` for the properties.
+- The new config interface approach: it uses **interfaces** annotated with `@io.smallrye.config.ConfigMapping` and `@io.quarkus.runtime.annotations.ConfigRoot` (for the config phase) and **methods** for the properties.
+
+The latter approach is what we have been recommending for quite some time now (it actually requires a special flag for the extension annotation processor to use the legacy config class approach).
+
+The old mechanism had a few issues that we wanted to fix, namely:
+
+- Classes were mutable, whereas Quarkus configuration is immutable, causing hard-to-debug problems and unexpected behaviours.
+- It only worked for extensions and in the context of Quarkus.
+
+SmallRye Config provides the currently recommended mechanism which can be used in any context (even without Quarkus). Using an interface instead of class makes it clear that the Config cannot be changed.
+On top of that, many of the new features were more straightforward to implement in this new model.
+
+There has been a massive effort to migrate from the former approach to the latter and we are happy to announce we migrated all the Core configuration to the `@ConfigMapping` approach.
+There are a few legacy classes left but they are there for compatibility purposes, for some config properties massively consumed outside of Quarkus that we didn't want to break (yet!).
+
+At this stage, it is time to start planning phasing out the legacy approach so that:
+
+- We avoid maintaining this code forever. It is complex and the bandwidth of the person maintaining it is limited.
+- We avoid some bugs that were never fixed in the config class approach due to either limited bandwidth or high complexity.
+- We reduce our startup footprint.
+
+By phasing out, we mean that the code won't be there anymore and that the approach with `@ConfigRoot` annotated **classes** won't be functional anymore, including for external extensions.
+
+The good thing is by migrating the Core configuration, we have loads of examples we can point people to and we are also very confident in the solidity of the `@ConfigMapping` code in SmallRye Config.
+
+Roberto has also led an effort to improve the performances of the new approach and allow it to scale better and results are promising (and still improving).
+
+The main goal of this ADR is to propose an approach for the transition that reduces friction.
+
+
+== Potential difficulties
+
+=== Transitioning
+
+Transitioning to config interfaces is a well known and straightforward process.
+It can be a bit tedious though for extensions with massive configuration or projects with a lot of extensions (Camel Quarkus comes to mind).
```

Comment:
```suggestion
It can be a bit tedious, though, for extensions with massive configurations or projects with a lot of extensions (Camel Quarkus comes to mind).
```